### PR TITLE
Update built dependency to 0.7.1

### LINKED ISSUE #451: RUSTSEC-2024-0013: Memory corruption, denial of service, and arbitrary code execution in libgit2
> Memory corruption, denial of service, and arbitrary code execution in libgit2

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Package             | `libgit2-sys`                      |
| Version             | `0.15.2+1.6.4`                   |
| URL                 | [https://github.com/rust-lang/git2-rs/pull/1017](https://github.com/rust-lang/git2-rs/pull/1017) |
| Date                | 2024-02-06                         |
| Patched versions    | `>=0.16.2`                  |

The [libgit2](https://github.com/libgit2/libgit2/) project fixed three security issues in the 1.7.2 release. These issues are:

* The `git_revparse_single` function can potentially enter an infinite loop on a well-crafted input, potentially causing a Denial of Service. This function is exposed in the `git2` crate via the [`Repository::revparse_single`](https://docs.rs/git2/latest/git2/struct.Repository.html#method.revparse_single) method.
* The `git_index_add` function may cause heap corruption and possibly lead to arbitrary code execution. This function is exposed in the `git2` crate via the [`Index::add`](https://docs.rs/git2/latest/git2/struct.Index.html#method.add) method.
* The smart transport negotiation may experience an out-of-bounds read when a remote server did not advertise capabilities.

The `libgit2-sys` crate bundles libgit2, or optionally links to a system libgit2 library. In either case, versions of the libgit2 library less than 1.7.2 are vulnerable. The 0.16.2 release of `libgit2-sys` bundles the fixed version of 1.7.2, and requires a system libgit2 version of at least 1.7.2.

It is recommended that all users upgrade.

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2024-0013.html) for additional details.

### PR DESCRIPTION
**Pull request by soenkeliebau**
# Description

Requires some small changes to the code, as the Option struct has been removed in favor of feature flags.

fixes #451 

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -27,5 +27,8 @@ strum = { version = "0.25", features = ["derive"] }
 tokio = { version = "1.29", features = ["full"] }
 tracing = "0.1"
 
+[patch.crates-io]
+built = { git = "https://github.com/stackabletech/built.git", branch = "fix/dont-write-env" }
```
Comment (nightkr) [inline Cargo.toml:31]: Do we want to submit a PR for this?

**Review by nightkr — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (soenkeliebau) [reply Cargo.toml]: I have opened https://github.com/lukaslueg/built/issues/62 to see if they'd be willing to consider one, but I would say in principle we would like to do that, yes.

**Comment by soenkeliebau** [comment]
Integration test still running, but looks okay so far I think ..

https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hbase-operator-it-custom/139/

**Comment by soenkeliebau** [comment]
Test was successful

**Review by nightkr — APPROVED** [review]

**Review by nightkr — APPROVED** [review]

**Comment by lfrancke** [comment]
I'm not sure I like this.
What does the SBOM for this look like? We now need to maintain this ourselves.

I want us to think twice before moving to a fork of something.

**Comment by lfrancke** [comment]
In fact: I think I'm in favor of reverting this.
We have a few more weeks to go and even then we can live with this vulnerability as it's nothing an end-user can control.

**Comment by soenkeliebau** [comment]
I am sure that I don't like it, but @nightkr and I discussed this and it seemed the "least bad" option that allows us to get rid of the RUSTSEC in our stack right now.

We can investigate moving of built onto another crate or wait if there is an appetite for our fix upstream. Both of which will take time which this workaround could buy us.

I am also happy to revert and wait for feedback on our issue upstream.

**Comment by lfrancke** [comment]
I would favor reverting just so everything is consistent but we can also keep it in and wait for upstream.
I don't want to maintain a fork. I think I'd rather figure out where the code is used and disable that part of `built` or instead basically say (in VEX terms) that the code cannot be controlled by users (which is the case as git should only be used during a github action run - which we control - to figure out the current commit/tag etc.)
So, I think we are not affected by this and I don't want to maintain a fork for something we don't need.

In other words: We can get rid of the RUSTSEC by just saying it's not valid here.

**Comment by soenkeliebau** [comment]
>  I think I'd rather figure out where the code is used and disable that part of built
That is exactly what the changed code in our fork does. The issue is, that upstream removed the ability to disable that code.

**Comment by soenkeliebau** [comment]
Actually, we only use built as a build dependency, so the code is not even present in our shipped artifacts, so in VEX terms we can even say "vulnerable code not present".

**Comment by NickLarsenNZ** [comment]
> Actually, we only use built as a build dependency, so the code is not even present in our shipped artifacts, so in VEX terms we can even say "vulnerable code not present".

Perhaps it is a non-issue in this case, but in general, even a build-time vulnerability can impact built artifacts. Just wanted to point it out so we don't get too comfortable ignoring build-time vulns.

**Comment by soenkeliebau** [comment]
Very true, and hopefully at some point it'll go away when we can upgrade again.

**Pull request merged**